Kernel/Arch/i386/Boot/boot.S

```diff
@@ -51,13 +51,82 @@ start:
     cli
     cld
 
+    # We first save the multiboot_info_ptr so it doesn't get trampled
+    addl $0xc0000000, %ebx
+    movl %ebx, multiboot_info_ptr - 0xc0000000
+
+    # First, let's set up the first page table to map the the first 4MiB of memory. 
+    # This makes sure we don't crash after we set CR3 and enable paging
+    pushl $0x200
+    movl $0x0, %ebx # ebx is the base pointer (kernel base is at physical address 0 in this case)
+    movl $((page_tables_start + (4096 * 1)) - 0xc0000000), %edx
+    call make_table
+
+    # Now we create the kernel mappings. The kernel maps 0MiB -> 8MiB into its address space at 
+    # v0xc0000000. 
+    pushl $0x400
+    movl $0x0, %ebx # ebx is the base pointer (kernel base is at physical address 0 in this case)
+    movl $((page_tables_start + (4096 * 2)) - 0xc0000000), %edx
+    call make_table
+
+    pushl $0x400
+    movl $0x400000, %ebx # ebx is the base pointer (kernel base is at physical address 0 in this case)
+    movl $((page_tables_start + (4096 * 3)) - 0xc0000000), %edx
+    call make_table
+        
+    
+    # Okay, so we have a page table that contains addresses of the first 4MiB of memory. Let's insert this into the 
+    # boot page directory. The index we need to insert it into is at vaddr >> 22, which is the page directory index.
+    # This reveals that we need to insert the page directory into 0xc0000000 >> 22 = 768
+    # An interesting quirk is that we must also identity map the first 4MiB too, as the next instruction after enabling
+    # paging is at a physical address, which cause a page fault. As we have no handler, this would cause a triple fault.
+    movl $((page_tables_start + (4096 * 1)) - 0xc0000000 + 0x003), page_tables_start - 0xc0000000 + 0
+    movl $((page_tables_start + (4096 * 2)) - 0xc0000000 + 0x003), page_tables_start - 0xc0000000 + 768 * 4
+    movl $((page_tables_start + (4096 * 3)) - 0xc0000000 + 0x003), page_tables_start - 0xc0000000 + 769 * 4
+
+    # Now let's load the CR3 register with our page directory
+    movl $(page_tables_start - 0xc0000000), %ecx
+    movl %ecx, %cr3
+
+    # Let's enable paging!
+    movl %cr0, %ecx
+    orl $0x80000001, %ecx
+	movl %ecx, %cr0 
+
+    lea high_address_space_start, %ecx
+    jmp *%ecx
+
+
+# Make a page table. This is called with the following arguments:
+# ebx = base pointer of mapping
+# edx = page table physical address
+#
+# Registers used in function
+# eax = loop counter
+# ecx = entry temp value
+make_table:
+    movl $0x00, %eax
```

Comment:
Bro-tip: `xor`-ing a register with itself to zero it is the oldest x86 micro-optimization the book :)
It generates a slightly shorter instruction since there's no need to encode an immediate value.

Reply:
> Bro-tip: xor-ing a register with itself to zero it is the oldest x86 micro-optimization the book :)
Argh! What a stupid thing to forget!